master/buildbot/master.py

```diff
@@ -334,53 +334,29 @@ def sendBuildbotNetUsageData(self):
 
     @defer.inlineCallbacks
     def stopService(self):
-        if self.masterid is not None:
-            yield self.data.updates.masterStopped(
-                name=self.name, masterid=self.masterid)
-        if self.running:
-            yield self.botmaster.cleanShutdown(quickMode=True, stopReactor=False)
-            yield service.AsyncMultiService.stopService(self)
-
-        log.msg("BuildMaster is stopped")
-        self._master_initialized = False
+        try:
+            yield self.initLock.acquire()
+            if self.masterid is not None:
+                yield self.data.updates.masterStopped(
+                    name=self.name, masterid=self.masterid)
+            if self.running:
+                yield self.botmaster.cleanShutdown(
+                    quickMode=True, stopReactor=False)
+                yield service.AsyncMultiService.stopService(self)
+
+            log.msg("BuildMaster is stopped")
+            self._master_initialized = False
+        finally:
+            yield self.initLock.release()
 
+    @defer.inlineCallbacks
     def reconfig(self):
-        # this method wraps doConfig, ensuring it is only ever called once at
-        # a time, and alerting the user if the reconfig takes too long
-        if self.reconfig_active:
-            log.msg("reconfig already active; will reconfig again after")
-            self.reconfig_requested = True
-            return
-
-        self.reconfig_active = self.reactor.seconds()
-        metrics.MetricCountEvent.log("loaded_config", 1)
-
-        # notify every 10 seconds that the reconfig is still going on, although
-        # reconfigs should not take that long!
-        self.reconfig_notifier = task.LoopingCall(lambda:
-                                                  log.msg("reconfig is ongoing for %d s" %
-                                                          (self.reactor.seconds() - self.reconfig_active)))
-        self.reconfig_notifier.start(10, now=False)
-
-        timer = metrics.Timer("BuildMaster.reconfig")
-        timer.start()
-
-        d = self.doReconfig()
-
-        @d.addBoth
-        def cleanup(res):
-            timer.stop()
-            self.reconfig_notifier.stop()
-            self.reconfig_notifier = None
-            self.reconfig_active = False
-            if self.reconfig_requested:
-                self.reconfig_requested = False
-                self.reconfig()
-            return res
-
-        d.addErrback(log.err, 'while reconfiguring')
-
-        return d  # for tests
+        try:
+            yield self.initLock.acquire()
+            metrics.MetricCountEvent.log("loaded_config", 1)
+            yield self.doReconfig()
+        finally:
+            yield self.initLock.release()
```

Comment:
maybe the better implementation is to just keep reconfig as is, and to change doReconfig to take the lock